Mage.Sets/src/mage/cards/f/FirestormPhoenix.java

```diff
@@ -0,0 +1,203 @@
+	    if (controller != null) {
+	        Permanent permanent = ((ZoneChangeEvent) event).getTarget();
+	        Cards revealedCards = new CardsImpl();
+	        game.getCard(permanent.getId());
+	        revealedCards.add(game.getCard(((ZoneChangeEvent) event).getTargetId()));
+	        if (permanent != null) {
+	        	
+	        	// Create ContinuousRuleModifyingEffect that restricts the casting of this card, until your next turn.
+	        	ContinuousEffect effectRestrict = new FirestormPhoenixRestrictEffect(permanent, game);
+	            game.addEffect(effectRestrict, source);
+	            
+	            // Create ContinuousEffectImpl that reveals FirestormPhoenix from hand continually, 
+	            // until it moves zone (is not in your hand) or until your next turn, whichever comes first. Need to add custom inActive.
+	            // Use game.informPlayers to inform players when the card changes zones.
+	            ContinuousEffect effectReveal = new FirestormPhoenixRevealEffect(controller, revealedCards, permanent.getIdName());
+	            game.addEffect(effectReveal, source);
+	        	
+	            // Move this card (FirestormPhoenix) to hand.
+	        	controller.moveCards(permanent, Zone.HAND, source, game);
+	        
+	            return true;
+	        }
+	    }
+	    return false;
+	}
+	
+    @Override
+    public boolean checksEventType(GameEvent event, Game game) {
+        return event.getType() == GameEvent.EventType.ZONE_CHANGE;
+    }
+	
+    @Override
+    public boolean applies(GameEvent event, Ability source, Game game) {
+    	return source.getSourceId().equals(event.getTargetId()) 
+        && ((ZoneChangeEvent) event).isDiesEvent();
+    }
+}
+
+class FirestormPhoenixRevealEffect extends ContinuousEffectImpl {
+	
+	private Player controller;
+	private Cards revealedCards;
+	private String cardName;
+
+	FirestormPhoenixRevealEffect(Player tmp_controller, Cards tmp_revealedCards, String tmp_cardName) {
+        super(Duration.UntilYourNextTurn, Outcome.Detriment);
+        controller = tmp_controller;
+        revealedCards = tmp_revealedCards;
+        cardName = tmp_cardName;
+        staticText = "Until that player’s next turn, that player plays with that card revealed in their hand and can’t play it";
+    }
+
+    private FirestormPhoenixRevealEffect(final FirestormPhoenixRevealEffect effect) {
+        super(effect);
+        this.controller = effect.controller;
+        this.revealedCards = effect.revealedCards;
+        this.cardName = effect.cardName;
+    }
+
+    @Override
+    public FirestormPhoenixRevealEffect copy() {
+        return new FirestormPhoenixRevealEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+    	return false;
+    }
+
+    @Override
+    public boolean apply(Layer layer, SubLayer sublayer, Ability source, Game game) {
+    	for(UUID permID : revealedCards) {
+    		if(game.getState().getZone(permID) == Zone.HAND) {
```

Comment:
not sufficient to check hand zone. need to check zcc (e.g. move to another zone and then back to hand is different object)

see above comment re: fixed target pointer

Reply:
Yeah, that's a good point; it was definitely overkill, but my understanding was it would be continually checking the game for any actions, so I didn't think there was a time when it could move to another zone without the effect being discarded. 

I'll make the changes so that it uses the zcc now.